server/src/main/java/com/microsoft/java/bs/core/internal/gradle/actions/GetSourceSetsAction.java

```diff
@@ -0,0 +1,110 @@
+package com.microsoft.java.bs.core.internal.gradle.actions;
+
+import com.microsoft.java.bs.gradle.model.BuildTargetDependency;
+import com.microsoft.java.bs.gradle.model.GradleSourceSet;
+import com.microsoft.java.bs.gradle.model.GradleSourceSets;
+import com.microsoft.java.bs.gradle.model.GradleSourceSetsMetadata;
+import com.microsoft.java.bs.gradle.model.impl.DefaultBuildTargetDependency;
+import com.microsoft.java.bs.gradle.model.impl.DefaultGradleSourceSet;
+import com.microsoft.java.bs.gradle.model.impl.DefaultGradleSourceSets;
+import org.gradle.tooling.BuildAction;
+import org.gradle.tooling.BuildController;
+import org.gradle.tooling.model.gradle.GradleBuild;
+
+import java.io.File;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * {@link BuildAction} that retrieves {@link DefaultGradleSourceSet} from a Gradle build,
+ * handling both normal and composite builds.
+ */
+public class GetSourceSetsAction implements BuildAction<GradleSourceSets> {
+
+  /**
+   * Executes the build action and retrieves source sets from the Gradle build.
+   *
+   * @return A {@link DefaultGradleSourceSets} object containing all retrieved source sets.
+   */
+  @Override
+  public GradleSourceSets execute(BuildController buildController) {
+    var traversedProjects = new ArrayList<String>();
+    var sourceSetToClasspath = new HashMap<GradleSourceSet, List<File>>();
+    var outputsToSourceSet = new HashMap<File, GradleSourceSet>();
+
+    GradleBuild buildModel = buildController.getBuildModel();
+    String rootProjectName = buildModel.getRootProject().getName();
+    fetchModels(buildController,
+        buildModel,
+        traversedProjects,
+        sourceSetToClasspath,
+        outputsToSourceSet,
+        rootProjectName);
+
+    // Add dependencies
+    var sourceSets = new ArrayList<GradleSourceSet>();
+    for (var entry : sourceSetToClasspath.entrySet()) {
+
+      var dependencies = new HashSet<BuildTargetDependency>();
+      for (File file : entry.getValue()) {
+        GradleSourceSet otherSourceSet = outputsToSourceSet.get(file);
+        if (otherSourceSet != null) {
```

Comment:
Is it possible that the otherSourceSet is the current sourceset being iterated?

I'm asking because I'm thinking that if it's necessary to check it here.

Reply:
I don't think a source set can have a dependency on it's own. Like a source set to output an artifact that exists in it's own compile classpath? It shouldn't be possible I believe. However I would let the Gradle community confirm it.

Reply:
At least it do no harm.

Reply:
Okay 👍